docs/architecture/adr-051-double-signing-protection-with-tendermint-mode.md

```diff
@@ -0,0 +1,104 @@
+    - Fullnode mode : fullnode mode does not have any capability to participate on consensus
+    - Validator mode : this mode is exactly same as existing state machine behavior. sync without voting on consensus, and participate consensus when fully synced
+    - Seed mode : lightweight seed mode only for maintain an address book, p2p like [TenderSeed](https://gitlab.com/polychainlabs/tenderseed)
+2. To provide a prevention method for double signing incidents mistakenly executed by validators
+    - Validators often mistakenly run duplicated validators to cause double-signing incident
+    - This proposed feature is to reduce the risk of mistaken double-signing incident by checking recent N blocks before voting begins
+    - When we think of such serious impact on double-signing incident, it is very reasonable to have multiple protection algorithm built in node daemon
+
+## Decision
+
+We would like to suggest a simple Tendermint mode abstraction with a double signing prevention method.
+
+### 1 ) Tendermint Mode
+
+- Which reactor, component to include for each node
+    - fullnode *(default)*
+        - switch, transport
+        - all reactors
+        - mempool, state
+        - rpc
+        - *~~no privValidator(priv_validator_key.json, priv_validator_state.json)~~*
+    - validator
+        - switch, transport
+        - all reactors
+        - mempool, state
+        - rpc
+        - with privValidator(priv_validator_key.json, priv_validator_state.json)
+    - seed
+        - switch, transport
+        - only pexReactor
+- Configuration, cli command
+    - We would like to suggest by introducing `mode` parameter in `config.toml` and cli
+    - `mode = "{{ .BaseConfig.Mode }}"` in `config.toml`
+    - `tendermint node --mode validator`  in cli
+    - fullnode | validator | seed (default: "fullnode")
+- RPC modification
+    - `host:26657/status`
+        - return empty `validator_info` when fullnode mode
+        - add return `mode` of this node
+    - no rpc server in seed mode
+- Where to modify in codebase
+    - Add  switch for `config.Mode` on `node/node.go:DefaultNewNode`
+    - If `config.Mode==validator`, call default `NewNode` (current logic)
+    - If `config.Mode==fullnode`, call `NewNode` with `nil` `privValidator` (do not load or generation)
+        - Need to add exception routine for `nil` `privValidator` to related functions
+    - If `config.Mode==seed`, call `NewSeedNode` (seed version of `node/node.go:NewNode`)
+        - Need to add exception routine for `nil` `reactor`, `component` to related functions
+
+### 2 ) Double signing prevention method
+
+- Methodology : query recent consensus results to find out whether node's consensus key is used on consensus recently or not
+- When to check
+    - When the state machine starts `ConsensusReactor` after fully synced
+    - When the node is validator ( with privValidator )
+    - When `cs.config.DoubleSignCheckHeight > 0`
```

Comment:
Can you explain why it has to be a parameter? I.e. why don't just use some constant like check 10 recent blocks?

Reply:
When the prevention method is on, restarting a validator node immediately will panic because the node itself voted on consensus with the same consensus key. So, validators should stop the state machine, wait for DoubleSignCheckHeight blocks and then restart the state machine to avoid panic stop.
I think some users don't want this side effect, so they want to turn it off or adjust the checking height as parameter,  they'll be able to turn it off by setting DoubleSignCheckHeight = 0

Reply:
> I think some users don't want this side effect

if it's only ON in `validator` mode, then choosing safety is fine in my opinion.

Reply:
yeah that could be good option, do you want specify that in ADR ?  or we could derive a conclusion on implementation PR process

Reply:
Yes, there definitely needs to be an option to disable this (like when doing enable active-active node coordination at the signer level).